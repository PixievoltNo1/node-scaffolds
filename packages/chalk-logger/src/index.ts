import type { Mutable } from '@guanghechen/utility-types'
import { partOptionsFromArgs } from './commander'
import type { Level } from './level'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export * from './color'
export * from './commander'
export * from './level'
export * from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = partOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flags: { ...options?.flags, ...optionsFromArgs.flags },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's level
   * @param level
   */
  public setLevel(level: Level | null | undefined): void {
    if (level == null) return
    const self = this as Mutable<this>
    self.level = level
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}

export default ChalkLogger
