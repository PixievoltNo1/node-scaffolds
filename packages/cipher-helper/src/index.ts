export * from './_base'
export * from './aes'
export * from './catalog'
export * from './types/catalog'
export * from './types/cipher-helper'
export * from './util/buffer'
export * from './util/key'
