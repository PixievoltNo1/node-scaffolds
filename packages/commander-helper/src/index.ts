export * from './command'
export * from './fs'
export * from './git'
export * from './merge'
export * from './option'
export * from './yarn'
