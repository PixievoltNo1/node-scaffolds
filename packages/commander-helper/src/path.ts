import path from 'path'

/**
 * Calc absolute path of p under the workspace
 *
 * @param workspace
 * @param targetPath
 */
export function absoluteOfWorkspace(
  workspace: string,
  targetPath?: string | null,
): string {
  if (targetPath == null) return workspace
  const absoluteDir: string = path.resolve(workspace, targetPath)
  return absoluteDir
}

/**
 * Calc relative path to workspace
 *
 * @param workspace
 * @param targetDir
 */
export function relativeOfWorkspace(
  workspace: string,
  targetDir: string,
): string {
  const absoluteDir = absoluteOfWorkspace(workspace, targetDir)
  const relativeDir: string = path.relative(workspace, absoluteDir)
  return path.normalize(relativeDir)
}
