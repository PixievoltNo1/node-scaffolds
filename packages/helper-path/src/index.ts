export * from './filepath'
export * from './urlpath'
