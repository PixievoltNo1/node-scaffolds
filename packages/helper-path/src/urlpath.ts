/**
 * Normalize url path.
 *
 * @param urlPath
 * @returns
 */
export const normalizeUrlPath = (urlPath: string): string => {
  const isAbsolute = urlPath.startsWith('/')
  const pathPieces = urlPath
    .split(/[/\\]+/g)
    .map(x => x.trim())
    .filter(piece => !!piece)
  const pieces: string[] = []
  for (const piece of pathPieces) {
    if (/^\./.test(piece)) {
      if (piece === '.') continue
      else if (piece === '..') {
        if (pieces.length > 0) {
          pieces.pop()
          continue
        } else if (isAbsolute) continue
      }
    }
    pieces.push(piece)
  }

  const p = pieces.join('/')
  return isAbsolute ? '/' + p : p || '.'
}
