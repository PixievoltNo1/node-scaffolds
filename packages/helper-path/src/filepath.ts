import path from 'node:path'

/**
 * Calc absolute path of p under the workspace.
 *
 * @param workspace
 * @param targetPath
 */
export function absoluteOfWorkspace(workspace: string, targetPath?: string | null): string {
  if (!targetPath) return workspace
  const filepath: string = path.resolve(workspace, targetPath)
  return path.normalize(filepath)
}

/**
 * Calc relative path to workspace.
 *
 * @param workspace
 * @param targetPath
 */
export function relativeOfWorkspace(workspace: string, targetPath: string): string {
  const absoluteDir = absoluteOfWorkspace(workspace, targetPath)
  const filepath: string = path.relative(workspace, absoluteDir)
  return path.normalize(filepath)
}
