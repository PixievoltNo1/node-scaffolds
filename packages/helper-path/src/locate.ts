import fs from 'fs'
import path from 'path'

/**
 * Locate a nearest filepath from the given `currentDir` which name included in the give `filenames`.
 *
 * @param currentDir
 * @param filenames
 * @returns
 */
export function locateNearestFilepath(
  currentDir: string,
  filenames: string | string[],
): string | null {
  // Uniform parameter format.
  if (!Array.isArray(filenames)) {
    // eslint-disable-next-line no-param-reassign
    filenames = [filenames]
  }

  for (const filename of filenames) {
    const filepath = path.join(currentDir, filename)
    if (fs.existsSync(filepath)) return filepath
  }

  const parentDir = path.dirname(currentDir)
  if (parentDir === currentDir) return null

  // Recursively locate.
  return locateNearestFilepath(parentDir, filenames)
}

/**
 * Find a nearest filepath from the give `currentDir`which matched the give tester `testFilepath`.
 *
 * @param currentDir
 * @param predicate
 * @returns
 */
export function findNearestFilepath(
  currentDir: string,
  predicate: (filepath: string) => boolean,
): string | null {
  const filenames = fs.readdirSync(currentDir)
  for (const filename of filenames) {
    const filepath = path.join(currentDir, filename)
    if (predicate(filepath)) return filepath
  }

  const parentDir = path.dirname(currentDir)
  if (parentDir === currentDir) return null

  // Recursively locate.
  return findNearestFilepath(parentDir, predicate)
}
