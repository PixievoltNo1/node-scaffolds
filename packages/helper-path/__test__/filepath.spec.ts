import { desensitize } from 'jest.helper'
import path from 'node:path'
import url from 'node:url'
import { absoluteOfWorkspace, relativeOfWorkspace } from '../src'

const __dirname = path.dirname(url.fileURLToPath(import.meta.url))

describe('absoluteOfWorkspace', function () {
  test('null / undefined', function () {
    expect(absoluteOfWorkspace(__dirname, null)).toEqual(__dirname)
    expect(absoluteOfWorkspace(__dirname, undefined)).toEqual(__dirname)
    expect(absoluteOfWorkspace(__dirname)).toEqual(__dirname)
  })

  test('basic', function () {
    expect(desensitize(absoluteOfWorkspace(__dirname, 'a/b/c.txt'))).toEqual(
      '<$WORKSPACE$>/packages/helper-path/__test__/a/b/c.txt',
    )
    expect(desensitize(absoluteOfWorkspace(__dirname, '/a/b/c.txt'))).toEqual('/a/b/c.txt')
    expect(desensitize(absoluteOfWorkspace(__dirname, path.join(__dirname, '/a/b/c.txt')))).toEqual(
      '<$WORKSPACE$>/packages/helper-path/__test__/a/b/c.txt',
    )
  })
})

describe('relativeOfWorkspace', function () {
  test('basic', function () {
    expect(desensitize(relativeOfWorkspace(__dirname, path.join(__dirname, 'a/b/c.txt')))).toEqual(
      'a/b/c.txt',
    )
    expect(
      desensitize(relativeOfWorkspace(__dirname, path.join(__dirname, '../a/b/c.txt'))),
    ).toEqual('../a/b/c.txt')
    expect(desensitize(relativeOfWorkspace(__dirname, path.join(__dirname, '../..')))).toEqual(
      '../..',
    )
  })
})
