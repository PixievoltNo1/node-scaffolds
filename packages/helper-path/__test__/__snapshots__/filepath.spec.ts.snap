// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mkdirsIfNotExists mkdirs logger 1`] = `
[
  [
    "[verb ] mkdirs: <$WORKSPACE$>/packages/helper-path/__test__/fixtures/basic--non-existed--2
",
  ],
]
`;
